Adds support for Graceful Updater in Compute Managed Instance Groups.

gsancewicz opened the pull request:
<!-- Put a description of what this PR is for here, along with any references to issues that this resolves or contributes to -->

Fixes https://github.com/hashicorp/terraform-provider-google/issues/9153

Adds support for Graceful Updater in Compute Managed Instance Groups.

Logic in previous [pull request](https://github.com/GoogleCloudPlatform/magic-modules/pull/4956) was not forcing sending empty fields and it was [reverted](https://github.com/GoogleCloudPlatform/magic-modules/pull/5964).

<!--
Replace each [ ] with [X] to check it. Switch to the preview view to make it easier to click on links.
These steps will speed up the review process, and we appreciate you spending time on them before sending
your code to be reviewed.
-->
If this PR is for Terraform, I acknowledge that I have:

- [X] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [X] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-downstream-tools), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [X] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/third_party/terraform/tests) (for handwritten resources or update tests).
- [ ] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/main/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
I was not able to run acceptance tests today since my local test environment is currently unavailable. I'm hoping that the pull request will trigger acceptance tests.
- [X] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

<!-- AUTOCHANGELOG for Downstream PRs.

Please select one of the following "release-note:" headings:
    - release-note:enhancement
    - release-note:bug
    - release-note:note
    - release-note:new-resource
    - release-note:new-datasource
    - release-note:deprecation
    - release-note:breaking-change
    - release-note:none
    
Unless you choose release-note:none, please add a release note.

See .ci/RELEASE_NOTES_GUIDE.md for writing good release notes.

You can add more release note blocks if you want more than one CHANGELOG
entry for this PR.
-->
**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
compute: added field `update_policy.most_disruptive_allowed_action` to `google_compute_instance_group_manager` and `google_compute_region_instance_group_manager`
```

```release-note:enhancement
compute: added value `REFRESH` to field update_policy.minimal_action` in `google_compute_instance_group_manager` and `google_compute_region_instance_group_manager`
```

modular-magician commented:
Hello!  I am a robot who works on Magic Modules PRs.

I've detected that you're a community contributor. @melinath, a repository maintainer, has been assigned to assist you and help review your changes. 

<details>
  <summary>:question: First time contributing? Click here for more details</summary>

---

Your assigned reviewer will help review your code by: 
* Ensuring it's backwards compatible, covers common error cases, etc.
* Summarizing the change into a user-facing changelog note.
* Passes tests, either our "VCR" suite, a set of presubmit tests, or with manual test runs.

You can help make sure that review is quick by running local tests and ensuring they're passing in between each push you make to your PR's branch. Also, try to leave a comment with each push you make, as pushes generally don't generate emails.

If your reviewer doesn't get back to you within a week after your most recent change, please feel free to leave a comment on the issue asking them to take a look! In the absence of a dedicated review dashboard most maintainers manage their pending reviews through email, and those will sometimes get lost in their inbox.

---

</details>

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-5975-old..auto-pr-5975) ( 6 files changed, 55 insertions(+), 30 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-5975-old..auto-pr-5975) ( 6 files changed, 57 insertions(+), 32 deletions(-))
TF Validator: [Diff](https://github.com/modular-magician/terraform-validator/compare/auto-pr-5975-old..auto-pr-5975) ( 2 files changed, 3 insertions(+), 3 deletions(-))

modular-magician commented:
#### Tests analytics 
Total tests: `1986` 
Passed tests `1744` 
Skipped tests: `240` 
Failed tests: `2` 

#### Action taken 
Triggering VCR tests in RECORDING mode for the following tests that failed during VCR: TestAccFirebaserulesRelease_BasicRelease|TestAccDatasourceGoogleServiceNetworkingPeeredDnsDomain_basic

modular-magician commented:
Tests passed during RECORDING mode:
 `TestAccDatasourceGoogleServiceNetworkingPeeredDnsDomain_basic`[[view](https://storage.cloud.google.com/vcr-test-logs/beta/refs/heads/auto-pr-5975/artifacts/e2273a44-2f2a-4955-b197-67d2b34032c5/recording/TestAccDatasourceGoogleServiceNetworkingPeeredDnsDomain_basic.log)]
`TestAccFirebaserulesRelease_BasicRelease`[[view](https://storage.cloud.google.com/vcr-test-logs/beta/refs/heads/auto-pr-5975/artifacts/e2273a44-2f2a-4955-b197-67d2b34032c5/recording/TestAccFirebaserulesRelease_BasicRelease.log)] 

All tests passed
View the [build log](https://storage.cloud.google.com/vcr-test-logs/beta/refs/heads/auto-pr-5975/artifacts/e2273a44-2f2a-4955-b197-67d2b34032c5/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/vcr-test-logs/beta/refs/heads/auto-pr-5975/artifacts/e2273a44-2f2a-4955-b197-67d2b34032c5/recording) for each test

gsancewicz commented:
/gcbrun

melinath commented:
tentatively reassigning to slevenick since it sounds like there may be context here. Feel free to assign back to me if needed.

slevenick commented on mmv1/third_party/terraform/tests/resource_compute_region_instance_group_manager_test.go.erb:
```diff
@@ -1295,15 +1295,15 @@ resource "google_compute_region_instance_group_manager" "igm-rolling-update-poli
   update_policy {
-    type                         = "PROACTIVE"
-    instance_redistribution_type = "NONE"
-    minimal_action               = "REPLACE"
```
Looks like minimal_action was removed here, but minimal_action is a required field for this resource, so the test cannot pass

gsancewicz replied:
I made this field optional with the default value REPLACE. This is the behavior of the REST api. minimal_action is not required but if it is not specified it will be set to REPLACE. I was able to execute tests this time using my environment and those tests passed there.

slevenick replied:
Why are we removing the Required parameter on this field? I believe we intentionally made this required to force users to specify REPLACE explicitly. Defaulting to REPLACE has somewhat dangerous implications so we try to make users make the choice themselves

gsancewicz replied:
I reverted this change. Lets decouple this discussion from this PR.

slevenick commented on mmv1/third_party/terraform/tests/resource_compute_instance_group_manager_test.go.erb:
```diff
@@ -1033,7 +1034,6 @@ resource "google_compute_instance_group_manager" "igm-rolling-update-policy" {
   target_size        = 3
   update_policy {
     type                  = "PROACTIVE"
-    minimal_action        = "REPLACE"
```
Minimal action needs to be added back here

gsancewicz replied:
ditto

gsancewicz replied:
I reverted this change. Lets decouple this discussion from this PR.

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-5975-old..auto-pr-5975) ( 6 files changed, 82 insertions(+), 34 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-5975-old..auto-pr-5975) ( 6 files changed, 83 insertions(+), 35 deletions(-))
TF Validator: [Diff](https://github.com/modular-magician/terraform-validator/compare/auto-pr-5975-old..auto-pr-5975) ( 2 files changed, 3 insertions(+), 3 deletions(-))

gsancewicz commented:
I added new commits with changes:
- changed ForceSendFields to NullFields in case when the field is not set. Difference in behavior is that if will send null instead of empty string
- minimal_action is not optional field with the default value REPLACE. REST Api behavior - minimal_action is optional but if it is not set it is set to REPLACE in the backend.

modular-magician commented:
#### Tests analytics 
Total tests: `1993` 
Passed tests `1749` 
Skipped tests: `241` 
Failed tests: `3` 

#### Action taken 
Triggering VCR tests in RECORDING mode for the following tests that failed during VCR: TestAccFirebaserulesRelease_BasicRelease|TestAccComputeRegionPerInstanceConfig_update|TestAccRegionInstanceGroupManager_stateful

slevenick commented on mmv1/third_party/terraform/website/docs/r/compute_instance_group_manager.html.markdown:
````diff
@@ -149,16 +149,19 @@ group. You can specify only one value. Structure is [documented below](#nested_a
 ```
 
-* `minimal_action` - (Required) - Minimal action to be taken on an instance. You can specify either `RESTART` to restart existing instances or `REPLACE` to delete and create new instances from the target template. If you specify a `RESTART`, the Updater will attempt to perform that action only. However, if the Updater determines that the minimal action you specify is not enough to perform the upd ...
+* `minimal_action` - (Required) - Minimal action to be taken on an instance. You can specify either `REFRESH` to update without stopping instances, `RESTART` to restart existing instances or `REPLACE` to delete and create new instances from the target template. If you specify a `REFRESH`, the Updater will attempt to perform that action only. However, if the Updater determines that the minimal act ...
````
The docs still state this is required

gsancewicz replied:
Fixed, Thank you for catching.

gsancewicz replied:
I reverted this change. Lets decouple this discussion from this PR.

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-5975-old..auto-pr-5975) ( 6 files changed, 82 insertions(+), 34 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-5975-old..auto-pr-5975) ( 6 files changed, 83 insertions(+), 35 deletions(-))
TF Validator: [Diff](https://github.com/modular-magician/terraform-validator/compare/auto-pr-5975-old..auto-pr-5975) ( 2 files changed, 3 insertions(+), 3 deletions(-))

modular-magician commented:
#### Tests analytics 
Total tests: `1995` 
Passed tests `1753` 
Skipped tests: `241` 
Failed tests: `1` 

#### Action taken 
Triggering VCR tests in RECORDING mode for the following tests that failed during VCR: TestAccFirebaserulesRelease_BasicRelease

modular-magician commented:
Tests passed during RECORDING mode:
 `TestAccFirebaserulesRelease_BasicRelease`[[view](https://storage.cloud.google.com/vcr-test-logs/beta/refs/heads/auto-pr-5975/artifacts/a589c4d3-fecb-45ea-85bf-b6575871b681/recording/TestAccFirebaserulesRelease_BasicRelease.log)] 

All tests passed
View the [build log](https://storage.cloud.google.com/vcr-test-logs/beta/refs/heads/auto-pr-5975/artifacts/a589c4d3-fecb-45ea-85bf-b6575871b681/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/vcr-test-logs/beta/refs/heads/auto-pr-5975/artifacts/a589c4d3-fecb-45ea-85bf-b6575871b681/recording) for each test

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-5975-old..auto-pr-5975) ( 6 files changed, 65 insertions(+), 29 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-5975-old..auto-pr-5975) ( 6 files changed, 66 insertions(+), 30 deletions(-))
TF Validator: [Diff](https://github.com/modular-magician/terraform-validator/compare/auto-pr-5975-old..auto-pr-5975) ( 2 files changed, 3 insertions(+), 3 deletions(-))

modular-magician commented:
#### Tests analytics 
Total tests: `2002` 
Passed tests `1758` 
Skipped tests: `242` 
Failed tests: `2` 

#### Action taken 
Triggering VCR tests in RECORDING mode for the following tests that failed during VCR: TestAccFirebaserulesRelease_BasicRelease|TestAccComputeForwardingRule_update

modular-magician commented:
Tests passed during RECORDING mode:
 `TestAccComputeForwardingRule_update`[[view](https://storage.cloud.google.com/vcr-test-logs/beta/refs/heads/auto-pr-5975/artifacts/d81dc690-8b6c-4cad-bc40-f7fce4f49f04/recording/TestAccComputeForwardingRule_update.log)]
`TestAccFirebaserulesRelease_BasicRelease`[[view](https://storage.cloud.google.com/vcr-test-logs/beta/refs/heads/auto-pr-5975/artifacts/d81dc690-8b6c-4cad-bc40-f7fce4f49f04/recording/TestAccFirebaserulesRelease_BasicRelease.log)] 

All tests passed
View the [build log](https://storage.cloud.google.com/vcr-test-logs/beta/refs/heads/auto-pr-5975/artifacts/d81dc690-8b6c-4cad-bc40-f7fce4f49f04/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/vcr-test-logs/beta/refs/heads/auto-pr-5975/artifacts/d81dc690-8b6c-4cad-bc40-f7fce4f49f04/recording) for each test
